xdsl/transforms/convert_riscv_to_llvm.py

```diff
@@ -0,0 +1,164 @@
+    op_type_rewrite_pattern,
+)
+from xdsl.traits import HasInsnRepresentation
+
+
+@dataclass(frozen=True)
+class RiscvToLLVMPattern(RewritePattern):
+    xlen: int
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: RISCVInstruction, rewriter: PatternRewriter):
+        ops_to_insert: list[Operation] = []
+
+        # inputs for the llvm inline asm op
+        assembly_args_str: list[str] = []
+        constraints: list[str] = []
+        inputs: list[SSAValue | OpResult] = []
+        # number of results produced from the inline assembly op
+        # all results are considered to be XLEN long integers
+        num_results: int = 0
+        # keep track which results are taken from the inline assembly op and which one are "other":
+        #  int      -> index into the inline asm op result list
+        #  ssa val  -> use this ssa value instead (e.g. when the op "returns" the zero register)
+        result_map: list[int | SSAValue] = []
+
+        # populate assembly_args_str and constraints
+        for arg in op.assembly_line_args():
+
+            # invalid argument
+            if not arg:
+                return
+
+            # ssa value used as an output operand
+            elif (
+                isinstance(arg, OpResult)
+                and isinstance(arg.type, IntRegisterType)
+                and arg.op is op
+            ):
+                # if we are storing to zero, we can't produce a result, so replace result by
+                # a get_register for the zero registers.
+                if arg.type.is_allocated and arg.type.index == IntAttr(0):
+                    assembly_args_str.append("x0")
+                    ops_to_insert.append(zero := riscv.GetRegisterOp(arg.type))
+                    # map final result to an existing SSA value
+                    result_map.append(zero.res)
+                    continue
+                # all other registers are treated as if they were unallocated
+                # meaning we cast them to i32 and pass values to the op
+                assembly_args_str.append(f"${len(inputs) + num_results}")
+                constraints.append("=r")
+                # map final result to a result of the inline asm op
+                result_map.append(num_results)
+                num_results += 1
+
+            # ssa value used as an input operand
+            elif isinstance(arg, SSAValue) and isinstance(arg.type, IntRegisterType):
+                # if the input is allocated to a register, use that register
+                if arg.type.is_allocated and arg.type.index == IntAttr(0):
+                    assembly_args_str.append("x0")
+                # otherwise we need to get the value from the SSA value
+                else:
+                    conversion_op = UnrealizedConversionCastOp.get([arg], [builtin.i32])
+                    ops_to_insert.append(conversion_op)
+                    inputs.append(conversion_op.outputs[0])
+                    constraints.append("r")
```

Comment:
I think we can use 'rI' here. LLVM is then smart enough to detect if it is possible to use an immediate, else it uses a register
```suggestion
                    constraints.append("rI")
```